doc/developers/advanced_installation.rst

```diff
@@ -96,7 +96,7 @@ feature, code or documentation improvement).
 
    .. prompt:: bash $
 
-     pip install -v --no-use-pep517 --no-build-isolation -e .
+     pip install -v -e .
```

Comment:
But the `--no-build-isolation` seems important, isn't it?

Otherwise, you are going to build with other version of library that are isolated and not use the one from your current environment.

If we go this way, we never need to request Cython for instance and only NumPy and SciPy to be installed for the runtime.

Reply:
I don't know either how build isolation works exactly. But if that's true maybe the doc shouldn't hint that `--no-build-isolation` is for saving time only then :thinking:

Reply:
Saving time for me it just a side effect of what is happening. The parameter allows you to:

- control the version of library used to compiled since you are not creating an isolated environment
- you will not recompile everything from scratch since you leverage the already compiled files since you are not isolated

Perfectly, we should not have to explain this in our documentation but have the proper explanation on the `pip` documentation.

I assume that the part about the PEP517 is something that we never really documented: https://github.com/scikit-learn/scikit-learn/pull/26334#issuecomment-1540062406

Reply:
Running twice `pip install -e . --no-build-isolation`, I can see that `pip` will create a tmp folder where to have the build and the compilation will happen at each call. Adding the `--no-use-pep-517` force to build into the `build` folder of the repository directly and it will not recompile if there is any change.

Reply:
The way I understand `--no-build-isolation` is as "use the libraries available in the environment in which `pip` is being run". It seems odd to me that you need to specify this, as an old person I think "what else would I want pip to do??", but times move on and I assume there is a good reason why the default isn't `--no-build-isolation`.

tl;dr: I think we should keep it, and we (as developers) understand exactly why it is there in the docs.

Until I read Guillaume's last comment here I was under the impression that no one fully understood what `--no-use-pep517` does. But from his comment it seems like he does. The naming is a bit weird, especially if you don't know your PEPs off by heart, but it seems it does something useful and we can explain it.

Should we explain it? I don't know, maybe link to the pip docs? 

As long as we don't know of any negative side-effects for the unsuspecting developer, I think we should keep the command as is (and feel sad that the names are a bit abstract). The windows build issue might be a reason to remove `--no-use-pep517`.

Reply:
> [build isolation](https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#build-isolation)
> For building packages using this interface, pip uses an _isolated environment_. That is, pip will install build-time Python dependencies in a temporary directory which will be added to `sys.path` for the build commands. This ensures that build requirements are handled independently of the user’s runtime environment.

Isn't there to understand that build dependencies are temporarily fetched into a separate, temporary file tree _if requirements are not already met by the user runtime environment_ ? I don't find any documentation of it actually having to do with compiled extensions and such.

> "what else would I want pip to do??"

I'd say that in general using different environments for different purposes is good, avoid a lot of troubles and unexpected interactions caused by unsuspected incompatible dependency trees and such, not having the build system alter the dependency of the environment that will be used at runtime is one of those instances of mixing things ~ especially for packages that can have build time dependencies that are not also runtime dependencies.